kong/timing/context.lua

```diff
@@ -0,0 +1,209 @@
+local cjson         = require("cjson.safe")
+local utils         = require("kong.tools.utils")
+
+local ngx_get_phase = ngx.get_phase
+local ngx_re_gmatch = ngx.re.gmatch
+
+local math_floor    = math.floor
+local setmetatable  = setmetatable
+local table_insert  = table.insert
+local table_remove  = table.remove
+
+local get_cur_msec  = utils.get_updated_monotonic_ms
+
+local assert        = assert
+
+local _M            = {}
+local _MT           = { __index = _M }
+
+
+function _M:enter_subcontext(name)
+  assert(name ~= nil, "name is required")
+  table_insert(self.sub_context_stack, self.current_subcontext)
+
+  if not self.current_subcontext.child then
+    self.current_subcontext.child = {}
+  end
+
+  if not self.current_subcontext.child[name] then
+    self.current_subcontext.child[name] = {}
+  end
+
+  self.current_subcontext = self.current_subcontext.child[name]
+  self.current_subcontext.____start____ = get_cur_msec()
+end
+
+
+function _M:leave_subcontext(attributes)
+  assert(#self.sub_context_stack > 0, "subcontext stack underflow")
+
+  local elapsed = get_cur_msec() - self.current_subcontext.____start____
+  local old_total_time = self.current_subcontext.total_time or 0
+  self.current_subcontext.total_time = math_floor(old_total_time + elapsed)
```

Comment:
Since we are using `monotoinc_msec()`, which is always a integer, so we should not use `math.floor` any more.

Reply:
fixed